src/metabase/automagic_dashboards/core.clj

```diff
@@ -438,149 +450,220 @@
-                                    ;; matching-rules returns an ArraySeq so first
-                                    ;; realises one element at a time (no chunking).
-                                    first))]
-     (let [indepth (->> rule
-                        :rule
-                        rules/indepth
-                        (keep (fn [indepth]
-                                (when-let [[dashboard _] (apply-rule root indepth)]
-                                  {:title       (:title dashboard)
-                                   :description (:description dashboard)
-                                   :table       (table root)
-                                   :url         (format "%s/%s/%s"
-                                                        (url root)
-                                                        (:rule rule)
-                                                        (:rule indepth))})))
-                        (take max-related))]
-       (log/info (format "Applying heuristic %s to %s."
-                         (:rule rule)
-                         (full-name root)))
-       (log/info (format "Dimensions bindings:\n%s"
-                         (->> dashboard
-                              :context
-                              :dimensions
-                              (m/map-vals #(update % :matches (partial map :name)))
-                              u/pprint-to-str)))
-       (log/info (format "Using definitions:\nMetrics:\n%s\nFilters:\n%s"
-                         (-> dashboard :context :metrics u/pprint-to-str)
-                         (-> dashboard :context :filters u/pprint-to-str)))
-       (-> dashboard
-           populate/create-dashboard
-           (assoc :related
-             {:tables  (->> root
-                            database
-                            candidate-tables
-                            (remove (comp #{root} :table))
-                            (take (- max-related (count indepth))))
-              :indepth indepth})))
-     (log/info (format "Skipping %s: no cards fully match bound dimensions."
-                       (full-name root))))))
+  [{:keys [rule show rules-prefix] :as root}]
+  (if-let [[dashboard rule]
+           (if rule
+             (apply-rule root rule)
+             (->> root
+                  (matching-rules (rules/load-rules rules-prefix))
+                  (keep (partial apply-rule root))
+                  ;; `matching-rules` returns an `ArraySeq` (via `sort-by`) so
+                  ;; `first` realises one element at a time (no chunking).
+                  first))]
+    (do
+      (log/info (format "Applying heuristic %s to %s."
+                        (:rule rule)
+                        (:full-name root)))
+      (log/info (format "Dimensions bindings:\n%s"
+                        (->> dashboard
+                             :context
+                             :dimensions
```

Comment:
I'm a little confused about the usage of dimensions throughout this code. What kind of structure is it? Here it looks like it's a map, in other areas it looks like it's a list of 1 keypair maps.

Reply:
In the heuristics it's a list of 1 keypair maps (to make the YAML at least somewhat bearable) with potential overloading. The overloading then gets resolved and dimensions get packed into a normal map. 
I'll think about better naming. Ideas welcome.